src/mjolnir/shortcutbuilder.cc

```diff
@@ -366,11 +365,18 @@ uint32_t AddShortcutEdges(GraphReader& reader,
                           const GraphId& start_node,
                           const uint32_t edge_index,
                           const uint32_t edge_count,
-                          std::unordered_map<uint32_t, uint32_t>& shortcuts) {
+                          std::unordered_map<uint32_t, uint32_t>& shortcuts,
+                          std::unordered_set<GraphId>& last_nodes) {
   // Shortcut edges have to start at a node that is not contracted - return if
-  // this node can be contracted.
+  // this node can be contracted and is not a terminus of a pervious shortcut.
+
+  // TODO: Could loop through last_nodes after proessing all tiles
+  // to continue the shortcut path of the ones terminated because they were
+  // too long. Currently, a shortcut path is continued by another tile at random.
   EdgePairs edgepairs;
-  if (CanContract(reader, tile, start_node, edgepairs)) {
+  std::unordered_set<GraphId>::const_iterator prev_shortcut = last_nodes.find(start_node);
+
+  if (CanContract(reader, tile, start_node, edgepairs) && prev_shortcut == last_nodes.end()) {
```

Comment:
is the idea with this change that we can try to continue making shortcuts for nodes that we aborted previous shortcuts at? basically we only bail on a node if its contractable but we havent ended/aborted a too-long shortcut at it? thus allowing the creation of another shortcut from that abort point? i just worry about the rest of the logic making sure not to duplicate the previously aborted shortcut and to continue in the other direction. are we sure this wouldnt cause any trouble?

Reply:
also, when creating a shortcut in one direction the code needs to make sure the "reverse" shortcut is created in the opposing direction. I didn't look at this closely, but making sure that rule is obeyed is essential.

Reply:
yeah exactly. basically changing anything about this is potentially brittle if we dont have a good understanding of the assumptions downstream. i have many worries like this in the back of my h ead as well. its not that i dont think we should try to tackle this but we need to mitigate any regressions this would cause. daves example is a good one. how do we make sure the opposing shortcut of the aborted one lines up where it stopped?